### PR TITLE
Track Form State in Fulcro

### PR DESCRIPTION
**Pull request by theianjones**
### Next Steps
- Add the mutation that submits the form to pathom
  - add the pathom mutation to send values to supabase
- add form validation too make sure all data inputs are entered

im trackin

### CODE REVIEW

File: shadow-cljs.edn [hunk 1]
```diff
@@ -17,15 +17,14 @@
                                                  holyjak.fulcro-troubleshooting ; optional, to find mistakes faster
                                                  com.fulcrologic.fulcro.inspect.preload
                                                  com.fulcrologic.fulcro.inspect.dom-picker-preload]
-                                    :http-root   "resources/public/" ;; Must be set to infer default purge targets.
-                                    :http-port   3000}
```
Comment (theianjones) [inline shadow-cljs.edn:21]: @zacjones93 you didnt need to add these values... they created 2 more builds for shadow cljs to track. We are using port 8000

File: src/com/example/client.cljs [hunk 2]
```diff
@@ -12,6 +12,9 @@
   (df/load! app :current-user ui/User {:target (targeting/replace-at [:root/user])
                                        :marker :load-user}))
 
+(defn load-readiness-quiz! [app]
+  (df/load! app :readiness-quiz ui/Quiz {:target (targeting/replace-at [:quiz])}))
```
Comment (theianjones) [inline src/com/example/client.cljs:16]: loading the global readiness quiz resolver from pathom into the fulcro db

File: src/com/example/data.cljs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+         :id "375623e0-92dd-4815-b507-4d577a55f37c"
+         :label "Readiness Quiz"
+         :questions
+         [#:question{:label "How many days in a row have you trained?"
+                     :id "trained",
+                     :options
+                     [#:option{:label "4+ days", :value 1}
+                      #:option{:label "3 days", :value 2}
+                      #:option{:label "2 days", :value 3}
+                      #:option{:label "1 days", :value 4}
```
Comment (theianjones) [inline src/com/example/data.cljs:14]: I updated this tree of data with namespaced keywords.

File: src/com/example/pathom.cljs [hunk 4]
```diff
@@ -33,15 +34,38 @@
...
+(pc/defresolver quiz-by-id [_ {id :quiz/id}]
+  {::pc/input #{:quiz/id}
+   ::pc/output [:quiz/id :quiz/version :quiz/label
+                {:quiz/questions [:question/id :question/label
+                                  {:question/options [:option/label :option/value]}]}]}
+  (quiz-by-id* id))
+
+(pc/defresolver readiness-quiz [_ _]
+  {::pc/output [{:readiness-quiz [:quiz/id]}]}
+  {:readiness-quiz {:quiz/id "375623e0-92dd-4815-b507-4d577a55f37c"}})
```
Comment (theianjones) [inline src/com/example/pathom.cljs:67]: this resolver just needs to send back an ident of `[:quiz/id id-value]` for pathom to connect it to the `quiz-by-id` resolver.

This enables queries for the quiz with this specific id without having to send the id in.

This is what the query looks like

```clojure
[{:readiness-quiz
  [:quiz/version
   :quiz/id
   :quiz/label
   {:quiz/questions
    [:question/id
     :question/label
     {:question/options [:option/label :option/value]}]}]}
 :com.wsscode.pathom.core/errors]
```

File: src/com/example/pathom.cljs [hunk 5]
```diff
@@ -66,7 +90,7 @@
 
 (def my-resolvers-and-mutations
   "Add any resolvers you make to this list (and reload to re-create the parser)"
-  [index-explorer create-random-thing i-fail person current-user user-authenticated?])
+  [index-explorer create-random-thing i-fail person current-user user-authenticated? quiz-by-id readiness-quiz])
```
Comment (theianjones) [inline src/com/example/pathom.cljs:93]: have to add our new resolvers to our pathom parser for pathom to know about the attributes it resolves.

File: src/com/example/pathom.cljs [hunk 6]
```diff
@@ -33,15 +34,38 @@
...
+                          :question/options (map (fn [o]
+                                                   {:option/label (:label o)
+                                                    :option/value (:value o)}) (:options q))}) questions)}))
+
+(pc/defresolver quiz-by-id [_ {id :quiz/id}]
+  {::pc/input #{:quiz/id}
+   ::pc/output [:quiz/id :quiz/version :quiz/label
+                {:quiz/questions [:question/id :question/label
+                                  {:question/options [:option/label :option/value]}]}]}
+  (quiz-by-id* id))
```
Comment (theianjones) [inline src/com/example/pathom.cljs:63]: `quiz-by-id*` is just a helper function that looks up the quiz in the `quizzes` array. The quiz object already has the correct keys for this query to resolve

File: src/com/example/ui.cljs [hunk 7]
```diff
@@ -30,24 +33,63 @@
...
+    (div
+     (dom/label {:htmlFor label} label)
+     (input-class input-props)
+     (dom/div {:classes [(when valid? "hidden")]}
+              error-message))))
+
+(defsc Question [this {:question/keys [id label options]}]
+  {:query [:question/id
+           :question/label
+           :ui/selected
```
Comment (theianjones) [inline src/com/example/ui.cljs:48]: `:ui/` namespaced keys are used only on the fulcro side. they will never be sent to pathom

File: src/com/example/ui.cljs [hunk 8]
```diff
@@ -30,24 +33,63 @@
...
+       (map (fn [o]
+              (field {:input-class  input
+                      :name         id
+                      :label        (:option/label o)
+                      :value        (str (:option/value o))
+                      :id           (:option/label o)
+                      :key          (:option/value o)
+                      :type         "radio"
+                      :autoComplete "off"
+                      :onChange     #(m/set-integer! this :ui/selected :event %)}))
```
Comment (theianjones) [inline src/com/example/ui.cljs:64]: when this field changes, we set the `ui/selected` field to the value that comes from the event object. `m/set-integer!` does the coercion to an int for us

File: src/com/example/ui.cljs [hunk 9]
```diff
@@ -30,24 +33,63 @@
...
+            options)))
+
+(def ui-question (comp/factory Question {:keyfn :question/id}))
+
+(defsc Quiz [this {:quiz/keys [version id label questions]}]
+  {:query [:quiz/version :quiz/id :quiz/label
+           {:quiz/questions (comp/get-query Question)}
+           fs/form-config-join]
```
Comment (theianjones) [inline src/com/example/ui.cljs:72]: eventually we can use the form state helpers but I dont want to implement undo yet.

File: src/com/example/mutations.cljs [hunk 10]
```diff
@@ -17,3 +19,26 @@
...
+  "Mutation helper: Updates state map with a cleared readiness form that is configured for form state support."
+  [state-map]
+  (-> state-map
+      (assoc-in readiness-ident
+                {:readiness/trained ""})
+      (fs/add-form-config* (readiness-class) readiness-ident)))
+
+(defmutation clear-readiness-form [_]
+  (action [{:keys [state]}]
+          (swap! state clear-readiness-form*)))
```
Comment (theianjones) [inline src/com/example/mutations.cljs:37]: these are unused. Keeping them around as an example

File: src/com/example/mutations.cljs [hunk 11]
```diff
@@ -17,3 +19,26 @@
...
+
+;; implement this when we want redo
+#_(defmutation use-quiz-form [{:quiz/keys [id]}]
```
Comment (theianjones) [inline src/com/example/mutations.cljs:40]: in fulcro, you send in a mutation to set up initial state for a form. then you can modify the values in your db with the guarantee that you can revert back to the original version.

Theres also a bunch of form validation stuff I havent looked into

File: src/com/example/data.cljs (hunk 3 above)
Comment (theianjones) [reply src/com/example/data.cljs:14]: This makes it so we can pass this object straight to our pathom resolver and not have to massage this data into what the resolver is looking for.

File: shadow-cljs.edn (hunk 1 above)
Comment (zacjones93) [reply shadow-cljs.edn:21]: ah, good to know

File: src/com/example/data.cljs (hunk 3 above)
Comment (zacjones93) [reply src/com/example/data.cljs:14]: where do you submit the form?

**Review by zacjones93 — APPROVED** [review]

**Pull request merged**